################################################ CALCULATE pij: PROBABILITY READ IS GENERATED BY CERTAIN SPECIES ###############################################Use Poisson probabilities to incorporate quality of match.
##The input file for this step is produced by the last step of the automated pipeline.
##It contains amino acid mismatches info based on blast output.
##Additionally info on how many reads were used to construct each contig.
##############################################################################################################################################################

library(Matrix)
library(data.table)

pipelineOut="output.txt"               ###produced by pipeline. TODO: pass it as an argument

allInfo<-read.table(file=pipelineOut, sep="\t", header=FALSE, quote="", comment.char="")

################### poisson prob with  lambda=9 mismatches per 100 a.a
### allInfo$V18 : number of amino acid mismatches.
### allInfo$V17: length of read
probPois<-ppois(allInfo$V18 - 1, lambda=0.09*((allInfo$V17)/3), lower.tail=FALSE)  ### at least that many mismatches could be observed by chance. Divide by 3 for translated read length
data_0.03<-cbind.data.frame(allInfo, probPois)                                 ### combine Poisson prob with other read info.
data_0.03$pij<-data_0.03$probPois/data_0.03$V21                        ### pij generative probability:  poisson prob over protein length 
data.dt<- data.table(data_0.03)
data.grouped<-data.dt[,list(max=max(pij)),by=c("V1", "V2", "V22", "V20")]  ### one hit per taxonid (best one)
pij<-as.data.frame(data.grouped)
colnames(pij)<-c("read","seq", "weight", "taxonID", "pij")           

####### read/contig weights 
readWeights<-unique(pij[,c("read", "weight")])
rownames(readWeights)<-readWeights[,"read"]

############use Sparse Matrix "reshape" long -> wide fromat
pijSparse<-with(pij, sparseMatrix(i = as.numeric(factor(read)), j=as.numeric((factor(taxonID))), x=pij,  dimnames=list(levels(factor(read)), levels(factor(taxonID)))))

pij.dt<-data.table(pij)
allspecies.dt<-pij.dt[,list(countReads=length(read)), by=list(taxonID)]
allspecies<-as.data.frame(allspecies.dt)
orderedSpecies<-allspecies[order(-allspecies[,2]) , ]                      #### order them by read count 

### add unknown bin and assign a pij
pijUnknown<-10^-20                  ### for BLASTn  / for BLASTx we use 10^-10 (depends on length of protein or genome respectively) 
pijSparseUnknown<-cBind(pijSparse, pijUnknown)
colnames(pijSparseUnknown)<-c(colnames(pijSparse), "unknown")

### sampling weight to use when choosing which species to add during MCMC
uniqReads<-dim(pijSparseUnknown)[[1]]
orderedSpecies$samplingWeight<-orderedSpecies[,2]/uniqReads

###Flattening the sampling probabilities
percentiles<-quantile(orderedSpecies$samplingWeight,  probs=c(0.2, 0.8))
orderedSpecies$samplingWeight[which(orderedSpecies$samplingWeight  >= percentiles["80%"])] <- percentiles["80%"]
orderedSpecies$samplingWeight[which(orderedSpecies$samplingWeight  <= percentiles["20%"])] <- percentiles["20%"]
colnames(orderedSpecies)<-c("taxonID", "countReads", "samplingWeight")

 
### remove objects
rm(list= ls()[!ls() %in% c("orderedSpecies","pijSparseUnknown", "readWeights")])
gc()


save.image("preprocess.RData")

